odoo/addons/base/models/qweb.py

```diff
@@ -397,397 +246,340 @@ def get_template(self, template, options):
+                            allowed_keys.append(tokens[i]['string'])
+                        i += 1
+
+            index += 1
+
+        # Use bracket to nest structures.
+        # This method is called recursively with the identified sections. During this recursive
+        # call we add to the allowed values the values provided by the list comprehension, lambda,
+        # etc., previously extracted.
+        index = 0
+        open_bracket_index = -1
+        bracket_depth = 0
+        while index < len(tokens):
+            t = tokens[index]
+            string = t['string']
+
+            if t['type'] == token.OP:
+                if string in ['(', '[', '{']:
+                    if bracket_depth == 0:
+                        open_bracket_index = index
+                    bracket_depth += 1
+                elif string in [')', ']', '}']:
+                    bracket_depth -= 1
+                    if bracket_depth == 0:
+                        code = self.__compile_expr_tokens(tokens[open_bracket_index + 1:index], list(allowed_keys), raise_on_missing)
+                        code = tokens[open_bracket_index]['string'] + code + tokens[index]['string']
+                        tokens[open_bracket_index:index + 1] = [{
+                            'string': code,
+                            'type': 'compiled',
+                            'start': tokens[open_bracket_index]['start'],
+                            'end': tokens[index]['end'],
+                        }]
+                        index = open_bracket_index
+
+            index += 1
+
+        # The keys will be namespaced by values if they are not allowed. In order to have a clear
+        # keyError message, this will be replaced by values['key'] for certain cases (for example
+        # if an attribute is called key.attrib, or an index key[0] ...)
+        code = []
+        index = 0
+        pos = tokens and tokens[0]['start'] # to keep indent when use expr on multi line
+        while index < len(tokens):
+            t = tokens[index]
+            string = t['string']
+
+            if t['start'][0] != pos[0]:
+                pos = (t['start'][0], 0)
+            space = t['start'][1] - pos[1]
+            if space:
+                code.append(' ' * space)
+            pos = t['start']
+
+            if t['type'] == token.NAME:
+                if string == 'lambda': # lambda => allowed values
+                    code.append('lambda ')
+                    index += 1
+                    while tokens[index] and (tokens[index]['type'] == token.OP or tokens[index]['type'] == token.NAME) and tokens[index]['string'] != ':':
+                        if tokens[index]['type'] == token.NAME:
+                            key = tokens[index]['string']
+                            code.append(key)
+                        else:
+                            code.append(tokens[index]['string'])
+                        index += 1
+                    code.append(':')
```

Comment:
`pos` needs to be adjusted here, otherwise the space-inserter gets lost and adds a bunch of intermediate space to the body e.g.

    lambda values: values

gets compiled to

    (lambda values:         values)